File: frontend/lib/models/vendor_settings.dart
/// DaySchedule represents operating hours for a single day
class DaySchedule {
  final String open; // "HH:MM" format (24-hour)
  final String close; // "HH:MM" format (24-hour)
  final bool closed; // If true, open/close times are ignored

  const DaySchedule({
    required this.open,
    required this.close,
    this.closed = false,
  });

  factory DaySchedule.fromJson(Map<String, dynamic> json) {
    return DaySchedule(
      open: json['open'] as String? ?? '09:00',
      close: json['close'] as String? ?? '21:00',
      closed: json['closed'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'open': open,
      'close': close,
      'closed': closed,
    };
  }

  DaySchedule copyWith({
    String? open,
    String? close,
    bool? closed,
  }) {
    return DaySchedule(
      open: open ?? this.open,
      close: close ?? this.close,
      closed: closed ?? this.closed,
    );
  }

  // Create a default "closed" schedule
  factory DaySchedule.closed() {
    return const DaySchedule(
      open: '09:00',
      close: '21:00',
      closed: true,
    );
  }

  // Create a default "open" schedule
  factory DaySchedule.open() {
    return const DaySchedule(
      open: '09:00',
      close: '21:00',
      closed: false,
    );
  }
}

/// HoursOfOperation represents weekly operating hours
class HoursOfOperation {
  final DaySchedule monday;
  final DaySchedule tuesday;
  final DaySchedule wednesday;
  final DaySchedule thursday;
  final DaySchedule friday;
  final DaySchedule saturday;
  final DaySchedule sunday;

  const HoursOfOperation({
    required this.monday,
    required this.tuesday,
    required this.wednesday,
    required this.thursday,
    required this.friday,
    required this.saturday,
    required this.sunday,
  });

  factory HoursOfOperation.fromJson(Map<String, dynamic> json) {
    return HoursOfOperation(
      monday: DaySchedule.fromJson(json['monday'] as Map<String, dynamic>? ?? {}),
      tuesday: DaySchedule.fromJson(json['tuesday'] as Map<String, dynamic>? ?? {}),
      wednesday: DaySchedule.fromJson(json['wednesday'] as Map<String, dynamic>? ?? {}),
      thursday: DaySchedule.fromJson(json['thursday'] as Map<String, dynamic>? ?? {}),
      friday: DaySchedule.fromJson(json['friday'] as Map<String, dynamic>? ?? {}),
      saturday: DaySchedule.fromJson(json['saturday'] as Map<String, dynamic>? ?? {}),
      sunday: DaySchedule.fromJson(json['sunday'] as Map<String, dynamic>? ?? {}),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'monday': monday.toJson(),
      'tuesday': tuesday.toJson(),
      'wednesday': wednesday.toJson(),
      'thursday': thursday.toJson(),
      'friday': friday.toJson(),
      'saturday': saturday.toJson(),
      'sunday': sunday.toJson(),
    };
  }

  // Create default hours (Mon-Sun 9:00-21:00)
  factory HoursOfOperation.defaultHours() {
    return HoursOfOperation(
      monday: DaySchedule.open(),
      tuesday: DaySchedule.open(),
      wednesday: DaySchedule.open(),
      thursday: DaySchedule.open(),
      friday: DaySchedule.open(),
      saturday: DaySchedule.open(),
      sunday: DaySchedule.open(),
    );
  }

  HoursOfOperation copyWith({
    DaySchedule? monday,
    DaySchedule? tuesday,
    DaySchedule? wednesday,
    DaySchedule? thursday,
    DaySchedule? friday,
    DaySchedule? saturday,
    DaySchedule? sunday,
  }) {
    return HoursOfOperation(
      monday: monday ?? this.monday,
      tuesday: tuesday ?? this.tuesday,
      wednesday: wednesday ?? this.wednesday,
      thursday: thursday ?? this.thursday,
      friday: friday ?? this.friday,
      saturday: saturday ?? this.saturday,
      sunday: sunday ?? this.sunday,
    );
  }

  // Get a day's schedule by name
  DaySchedule getDaySchedule(String dayName) {
    switch (dayName.toLowerCase()) {
      case 'monday':
        return monday;
      case 'tuesday':
        return tuesday;
      case 'wednesday':
        return wednesday;
      case 'thursday':
        return thursday;
      case 'friday':
        return friday;
      case 'saturday':
        return saturday;
      case 'sunday':
        return sunday;
      default:
        return DaySchedule.open();
    }
  }
}

/// VendorSettings represents vendor restaurant settings for a specific restaurant
class VendorSettings {
  final int restaurantId;
  final String restaurantName;
  final int averagePrepTimeMinutes;
  final HoursOfOperation? hoursOfOperation;

  const VendorSettings({
    required this.restaurantId,
    required this.restaurantName,
    required this.averagePrepTimeMinutes,
    this.hoursOfOperation,
  });

  factory VendorSettings.fromJson(Map<String, dynamic> json) {
    return VendorSettings(
      restaurantId: json['restaurant_id'] as int,
      restaurantName: json['restaurant_name'] as String,
      averagePrepTimeMinutes: json['average_prep_time_minutes'] as int,
      hoursOfOperation: json['hours_of_operation'] != null
          ? HoursOfOperation.fromJson(json['hours_of_operation'] as Map<String, dynamic>)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'restaurant_id': restaurantId,
      'restaurant_name': restaurantName,
      'average_prep_time_minutes': averagePrepTimeMinutes,
      if (hoursOfOperation != null) 'hours_of_operation': hoursOfOperation!.toJson(),
    };
  }

  VendorSettings copyWith({
    int? restaurantId,
    String? restaurantName,
    int? averagePrepTimeMinutes,
    HoursOfOperation? hoursOfOperation,
  }) {
    return VendorSettings(
      restaurantId: restaurantId ?? this.restaurantId,
      restaurantName: restaurantName ?? this.restaurantName,
      averagePrepTimeMinutes: averagePrepTimeMinutes ?? this.averagePrepTimeMinutes,
      hoursOfOperation: hoursOfOperation ?? this.hoursOfOperation,
    );
  }
}

/// UpdateVendorSettingsRequest represents the request to update vendor restaurant settings
class UpdateVendorSettingsRequest {
  final int? averagePrepTimeMinutes;
  final HoursOfOperation? hoursOfOperation;

  const UpdateVendorSettingsRequest({
    this.averagePrepTimeMinutes,
    this.hoursOfOperation,
  });

  Map<String, dynamic> toJson() {
    return {
      if (averagePrepTimeMinutes != null)
        'average_prep_time_minutes': averagePrepTimeMinutes,
      if (hoursOfOperation != null)
        'hours_of_operation': hoursOfOperation!.toJson(),
    };
  }

  // Validation helper
  bool get hasUpdates =>
      averagePrepTimeMinutes != null || hoursOfOperation != null;
}

File: backend/handlers/vendor_settings.go
package handlers

import (
	"delivery_app/backend/middleware"
	"delivery_app/backend/models"
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/gorilla/mux"
)

// GetRestaurantSettings retrieves restaurant settings for the authenticated vendor or admin
// GET /api/vendor/restaurant/{restaurantId}/settings
func (h *Handler) GetRestaurantSettings(w http.ResponseWriter, r *http.Request) {
	// Get authenticated user
	user := middleware.MustGetUserFromContext(r.Context())

	// Get restaurant ID from URL
	vars := mux.Vars(r)
	restaurantIDStr := vars["restaurantId"]
	restaurantID, err := strconv.Atoi(restaurantIDStr)
	if err != nil {
		sendError(w, http.StatusBadRequest, "Invalid restaurant ID")
		return
	}

	// Authorization check
	var restaurant *models.Restaurant
	if user.UserType == models.UserTypeAdmin {
		// Admin can access any restaurant
		restaurant, err = h.App.Deps.Restaurants.GetByID(restaurantID)
		if err != nil {
			sendError(w, http.StatusNotFound, "Restaurant not found")
			return
		}
	} else if user.UserType == models.UserTypeVendor {
		// Vendor can only access their own restaurants
		vendorID, err := h.App.Deps.Users.GetVendorIDByUserID(user.UserID)
		if err != nil {
			sendError(w, http.StatusForbidden, "Vendor profile not found")
			return
		}

		// Verify ownership and get restaurant
		restaurant, err = h.App.Deps.Restaurants.GetByIDWithOwnershipCheck(restaurantID, vendorID)
		if err != nil {
			if strings.Contains(err.Error(), "does not belong") {
				sendError(w, http.StatusForbidden, "You do not have permission to access this restaurant")
				return
			}
			sendError(w, http.StatusNotFound, "Restaurant not found")
			return
		}
	} else {
		sendError(w, http.StatusForbidden, "Only vendors and admins can access restaurant settings")
		return
	}

	// Parse hours of operation if present
	var hoursOfOperation *models.HoursOfOperation
	if restaurant.HoursOfOperation != nil && *restaurant.HoursOfOperation != "" {
		hoursOfOperation = &models.HoursOfOperation{}
		if err := json.Unmarshal([]byte(*restaurant.HoursOfOperation), hoursOfOperation); err != nil {
			// If parsing fails, return nil for hours
			hoursOfOperation = nil
		}
	}

	// Build response
	settings := models.VendorSettings{
		RestaurantID:       restaurant.ID,
		RestaurantName:     restaurant.Name,
		AveragePrepTimeMin: restaurant.AveragePrepTimeMin,
		HoursOfOperation:   hoursOfOperation,
	}

	sendJSON(w, http.StatusOK, map[string]interface{}{
		"success": true,
		"data":    settings,
	})
}

// UpdateRestaurantSettings updates restaurant settings (hours and/or prep time)
// PUT /api/vendor/restaurant/{restaurantId}/settings
func (h *Handler) UpdateRestaurantSettings(w http.ResponseWriter, r *http.Request) {
	// Get authenticated user
	user := middleware.MustGetUserFromContext(r.Context())

	// Get restaurant ID from URL
	vars := mux.Vars(r)
	restaurantIDStr := vars["restaurantId"]
	restaurantID, err := strconv.Atoi(restaurantIDStr)
	if err != nil {
		sendError(w, http.StatusBadRequest, "Invalid restaurant ID")
		return
	}

	// Authorization check
	if user.UserType == models.UserTypeAdmin {
		// Admin can modify any restaurant - verify it exists
		_, err := h.App.Deps.Restaurants.GetByID(restaurantID)
		if err != nil {
			sendError(w, http.StatusNotFound, "Restaurant not found")
			return
		}
	} else if user.UserType == models.UserTypeVendor {
		// Vendor can only modify their own restaurants
		vendorID, err := h.App.Deps.Users.GetVendorIDByUserID(user.UserID)
		if err != nil {
			sendError(w, http.StatusForbidden, "Vendor profile not found")
			return
		}

		// Verify ownership
		if err := h.App.Deps.Restaurants.VerifyVendorOwnership(restaurantID, vendorID); err != nil {
			if strings.Contains(err.Error(), "does not belong") {
				sendError(w, http.StatusForbidden, "You do not have permission to modify this restaurant")
				return
			}
			sendError(w, http.StatusNotFound, "Restaurant not found")
			return
		}
	} else {
		sendError(w, http.StatusForbidden, "Only vendors and admins can modify restaurant settings")
		return
	}

	// Decode request
	var req models.UpdateVendorSettingsRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body")
		return
	}

	// Validate at least one field is provided
	if req.AveragePrepTimeMin == nil && req.HoursOfOperation == nil {
		sendError(w, http.StatusBadRequest, "At least one field must be provided")
		return
	}

	// Validate prep time if provided
	if req.AveragePrepTimeMin != nil {
		if *req.AveragePrepTimeMin < 1 || *req.AveragePrepTimeMin > 300 {
			sendError(w, http.StatusBadRequest, "Average prep time must be between 1 and 300 minutes")
			return
		}
	}

	// Validate hours of operation if provided
	if req.HoursOfOperation != nil {
		if err := validateHoursOfOperation(req.HoursOfOperation); err != nil {
			sendError(w, http.StatusBadRequest, fmt.Sprintf("Invalid hours of operation: %s", err.Error()))
			return
		}
	}

	// Convert hours to JSON string if provided
	var hoursJSON *string
	if req.HoursOfOperation != nil {
		hoursBytes, err := json.Marshal(req.HoursOfOperation)
		if err != nil {
			sendError(w, http.StatusInternalServerError, "Failed to process hours of operation")
			return
		}
		hoursStr := string(hoursBytes)
		hoursJSON = &hoursStr
	}

	// Update restaurant settings
	if err := h.App.Deps.Restaurants.UpdateRestaurantSettings(restaurantID, hoursJSON, req.AveragePrepTimeMin); err != nil {
		sendErrorWithContext(w, r, http.StatusInternalServerError, "Failed to update restaurant settings", err)
		return
	}

	// Get updated restaurant
	restaurant, err := h.App.Deps.Restaurants.GetByID(restaurantID)
	if err != nil {
		sendError(w, http.StatusInternalServerError, "Failed to retrieve updated restaurant")
		return
	}

	// Parse hours of operation
	var hoursOfOperation *models.HoursOfOperation
	if restaurant.HoursOfOperation != nil && *restaurant.HoursOfOperation != "" {
		hoursOfOperation = &models.HoursOfOperation{}
		if err := json.Unmarshal([]byte(*restaurant.HoursOfOperation), hoursOfOperation); err != nil {
			hoursOfOperation = nil
		}
	}

	// Build response
	settings := models.VendorSettings{
		RestaurantID:       restaurant.ID,
		RestaurantName:     restaurant.Name,
		AveragePrepTimeMin: restaurant.AveragePrepTimeMin,
		HoursOfOperation:   hoursOfOperation,
	}

	sendJSON(w, http.StatusOK, map[string]interface{}{
		"success": true,
		"message": "Restaurant settings updated successfully",
		"data":    settings,
	})
}

// UpdateRestaurantPrepTime updates only the average prep time
// PATCH /api/vendor/restaurant/{restaurantId}/prep-time
func (h *Handler) UpdateRestaurantPrepTime(w http.ResponseWriter, r *http.Request) {
	// Get authenticated user
	user := middleware.MustGetUserFromContext(r.Context())

	// Get restaurant ID from URL
	vars := mux.Vars(r)
	restaurantIDStr := vars["restaurantId"]
	restaurantID, err := strconv.Atoi(restaurantIDStr)
	if err != nil {
		sendError(w, http.StatusBadRequest, "Invalid restaurant ID")
		return
	}

	// Authorization check
	if user.UserType == models.UserTypeAdmin {
		// Admin can modify any restaurant - verify it exists
		_, err := h.App.Deps.Restaurants.GetByID(restaurantID)
		if err != nil {
			sendError(w, http.StatusNotFound, "Restaurant not found")
			return
		}
	} else if user.UserType == models.UserTypeVendor {
		// Vendor can only modify their own restaurants
		vendorID, err := h.App.Deps.Users.GetVendorIDByUserID(user.UserID)
		if err != nil {
			sendError(w, http.StatusForbidden, "Vendor profile not found")
			return
		}

		// Verify ownership
		if err := h.App.Deps.Restaurants.VerifyVendorOwnership(restaurantID, vendorID); err != nil {
			if strings.Contains(err.Error(), "does not belong") {
				sendError(w, http.StatusForbidden, "You do not have permission to modify this restaurant")
				return
			}
			sendError(w, http.StatusNotFound, "Restaurant not found")
			return
		}
	} else {
		sendError(w, http.StatusForbidden, "Only vendors and admins can modify restaurant settings")
		return
	}

	// Decode request
	var req models.UpdatePrepTimeRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		sendError(w, http.StatusBadRequest, "Invalid request body")
		return
	}

	// Validate prep time
	if req.AveragePrepTimeMin < 1 || req.AveragePrepTimeMin > 300 {
		sendError(w, http.StatusBadRequest, "Average prep time must be between 1 and 300 minutes")
		return
	}

	// Update prep time
	if err := h.App.Deps.Restaurants.UpdateRestaurantPrepTime(restaurantID, req.AveragePrepTimeMin); err != nil {
		sendErrorWithContext(w, r, http.StatusInternalServerError, "Failed to update prep time", err)
		return
	}

	sendJSON(w, http.StatusOK, map[string]interface{}{
		"success": true,
		"message": "Average prep time updated successfully",
		"data": map[string]interface{}{
			"restaurant_id":             restaurantID,
			"average_prep_time_minutes": req.AveragePrepTimeMin,
		},
	})
}

// validateHoursOfOperation validates the hours of operation structure
func validateHoursOfOperation(hours *models.HoursOfOperation) error {
	if hours == nil {
		return fmt.Errorf("hours of operation cannot be nil")
	}

	days := []struct {
		name     string
		schedule models.DaySchedule
	}{
		{"Monday", hours.Monday},
		{"Tuesday", hours.Tuesday},
		{"Wednesday", hours.Wednesday},
		{"Thursday", hours.Thursday},
		{"Friday", hours.Friday},
		{"Saturday", hours.Saturday},
		{"Sunday", hours.Sunday},
	}

	for _, day := range days {
		if !day.schedule.Closed {
			// Validate time format (HH:MM)
			if err := validateTimeFormat(day.schedule.Open); err != nil {
				return fmt.Errorf("%s open time: %w", day.name, err)
			}
			if err := validateTimeFormat(day.schedule.Close); err != nil {
				return fmt.Errorf("%s close time: %w", day.name, err)
			}

			// Validate open < close
			openTime, _ := time.Parse("15:04", day.schedule.Open)
			closeTime, _ := time.Parse("15:04", day.schedule.Close)
			if !openTime.Before(closeTime) {
				return fmt.Errorf("%s: open time must be before close time", day.name)
			}
		}
	}

	return nil
}

// validateTimeFormat validates time is in HH:MM format
func validateTimeFormat(timeStr string) error {
	_, err := time.Parse("15:04", timeStr)
	if err != nil {
		return fmt.Errorf("invalid time format, must be HH:MM (24-hour)")
	}
	return nil
}
